npm_and_yarn/spec/dependabot/npm_and_yarn/helpers_spec.rb

```diff
@@ -501,4 +501,52 @@
       expect(described_class.node_version).to be_nil
     end
   end
+
+  describe "::find_cached_version" do
+    before do
+      # Create a temporary cache directory structure for testing
+      FileUtils.mkdir_p("/tmp/test_cache/npm")
+      FileUtils.touch("/tmp/test_cache/npm/11.6.2")
+      FileUtils.touch("/tmp/test_cache/npm/10.9.4")
+      FileUtils.touch("/tmp/test_cache/npm/9.9.4")
+    end
+
+    after do
+      FileUtils.rm_rf("/tmp/test_cache")
+    end
+
+    it "finds exact version match" do
+      allow(Dir).to receive(:exist?).with("/home/dependabot/.cache/node/corepack/v1/npm").and_return(true)
+      allow(Dir).to receive(:entries).with("/home/dependabot/.cache/node/corepack/v1/npm")
+                                     .and_return([".", "..", "11.6.2", "10.9.4", "9.9.4"])
+
+      result = described_class.find_cached_version("npm", "11.6.2")
+      expect(result).to eq("11.6.2")
+    end
+
+    it "finds highest matching version for major version request" do
+      allow(Dir).to receive(:exist?).with("/home/dependabot/.cache/node/corepack/v1/npm").and_return(true)
+      allow(Dir).to receive(:entries).with("/home/dependabot/.cache/node/corepack/v1/npm")
+                                     .and_return([".", "..", "11.6.2", "10.9.4", "9.9.4"])
+
+      result = described_class.find_cached_version("npm", "11")
+      expect(result).to eq("11.6.2")
+    end
+
+    it "returns nil when cache directory doesn't exist" do
+      allow(Dir).to receive(:exist?).with("/home/dependabot/.cache/node/corepack/v1/npm").and_return(false)
+
+      result = described_class.find_cached_version("npm", "11")
+      expect(result).to be_nil
+    end
```

Comment:
I do not believe that this can happen unless there is something wrong with our corepack installation